stdlib/list.ml

```diff
@@ -77,6 +77,18 @@ let rec flatten = function
 
 let concat = flatten
 
+let[@tail_mod_cons] rec intersperse sep = function
+  | ([] | [ _ ]) as l -> l
+  | a :: (_ :: _ as l) -> a :: sep :: intersperse sep l
+
+let intercalate sep =
+  let rec aux acc = function
+    | [] | [[]] -> rev acc
+    | [] :: (_ :: _ as l) -> aux (rev_append sep acc) l
+    | (a :: l1) :: l2 -> aux (a :: acc) (l1 :: l2)
+  in
+  aux []
```

Comment:
The following TRMC implementation is always(?) faster than (and sometimes twice as fast as) the accumulate-and-reverse implementation.

```suggestion
let[@tail_mod_cons] rec intercalate sep = function
  | [] -> []
  | xs :: [] -> xs
  | xs :: (_ :: _ as ls) -> ic_one sep ls xs
and[@tail_mod_cons] ic_one sep ls = function
  | [] -> ic_sep sep ls sep
  | x :: xs -> x :: ic_one sep ls xs
and[@tail_mod_cons] ic_sep sep ls = function
  | [] -> intercalate sep ls
  | x :: xs -> x :: ic_sep sep ls xs
```

Reply:
In your versions, I underestand that `xs` is made the last argument to use the `function` form, but I would prefer if those functions were written with `xs ls` as parameters instead of `ls xs`, to respect the spatial ordering of elements in the input and output.

Reply:
I appreciate the suggestion (and its justification), but I prefer to keep the code the way it's written.  The syntactic overhead of `match` is quite significant in this case, since the functions are so short and simple.

Reply:
Don't you think that the functions `ic_one` and `ic_sep` should be defined inside `intercalate` itself, not to pollute the global scope?

Reply:
Defining them at module level is fine, because the module interface ensures that they're not visible externally.

Reply:
I've just pushed exactly as @yallop suggested.